core/dns_test.go

```diff
@@ -14,8 +17,97 @@ import (
 	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/miekg/dns"
 )
 
+const dnsLoopbackAddr = "127.0.0.1:4053"
+
+func mockDNSQuery(w dns.ResponseWriter, r *dns.Msg) {
+	defer w.Close()
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Compress = false
+
+	for _, q := range r.Question {
+		switch q.Qtype {
+		case dns.TypeSOA:
+			record := new(dns.SOA)
+			record.Hdr = dns.RR_Header{Name: "lets-encrypt.org.", Rrtype: dns.TypeSOA, Class: dns.ClassINET, Ttl: 0}
+			record.Ns = "ns.lets-encrypt.org."
+			record.Mbox = "master.lets-encrypt.org."
+			record.Serial = 1
+			record.Refresh = 1
+			record.Retry = 1
+			record.Expire = 1
+			record.Minttl = 1
+
+			m.Answer = append(m.Answer, record)
+			w.WriteMsg(m)
+			return
+		case dns.TypeA:
+			switch q.Name {
+			case "cps.letsencrypt.org.":
+				record := new(dns.A)
+				record.Hdr = dns.RR_Header{Name: "cps.lets-encrypt.org.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0}
+				record.A = net.ParseIP("127.0.0.1")
+
+				m.Answer = append(m.Answer, record)
+				w.WriteMsg(m)
+				return
+			case "sigfail.verteiltesysteme.net.":
+				if !r.CheckingDisabled {
+					m.Rcode = dns.RcodeServerFailure
+				}
+				w.WriteMsg(m)
+				return
+			}
+		case dns.TypeCAA:
+			record := new(dns.CAA)
+			record.Hdr = dns.RR_Header{Name: "bracewel.net.", Rrtype: dns.TypeCAA, Class: dns.ClassINET, Ttl: 0}
+			record.Tag = "issue"
+			record.Value = "letsencrypt.org"
+			record.Flag = 1
+
+			m.Answer = append(m.Answer, record)
+			w.WriteMsg(m)
+			return
+		}
+	}
+
+	w.WriteMsg(m)
+	return
+}
+
+func serveLoopResolver(stopChan chan bool) chan bool {
+	dns.HandleFunc(".", mockDNSQuery)
+	server := &dns.Server{Addr: dnsLoopbackAddr, Net: "udp", ReadTimeout: time.Millisecond, WriteTimeout: time.Millisecond}
+	waitChan := make(chan bool, 1)
+	go func() {
+		waitChan <- true
+		err := server.ListenAndServe()
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+	}()
+	go func() {
+		<-stopChan
+		err := server.Shutdown()
+		if err != nil {
+			fmt.Println(err)
+		}
+	}()
+	return waitChan
+}
+
+func TestMain(m *testing.M) {
```

Comment:
For anyone who isn't familiar, as I wasn't, this is the supported way of doing "BeforeClass" / "AfterClass" style functionality: https://golang.org/pkg/testing/#hdr-Main